core/src/unprocessed_packet_batches.rs

```diff
@@ -245,12 +250,43 @@ impl UnprocessedPacketBatches {
         self.message_hash_to_transaction.iter_mut().map(|(_k, v)| v)
     }
 
+    /// iterates DesserializedPackets in descending priority (max-first) order,
+    /// calls FnMut for each DeserializedPacket.
+    pub fn iter_desc<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&mut DeserializedPacket),
+    {
+        let mut packet_priority_queue: MinMaxHeap<Rc<ImmutableDeserializedPacket>> = self
+            .packet_priority_queue
+            .drain_desc()
+            .map(|immutable_packet| {
+                match self
+                    .message_hash_to_transaction
+                    .entry(*immutable_packet.message_hash())
+                {
+                    Entry::Vacant(_vacant_entry) => {
+                        panic!(
+                            "entry {} must exist to be consistent with `packet_priority_queue`",
+                            immutable_packet.message_hash()
+                        );
+                    }
+                    Entry::Occupied(mut occupied_entry) => {
+                        f(occupied_entry.get_mut());
+                        immutable_packet
+                    }
+                }
+            })
+            .collect();
```

Comment:
Another thought, I feel like the `fill_batches` here: https://github.com/solana-labs/solana/blob/7dc554bbe398eac0e7e8094ffca48c410df6e6f2/core/src/forward_packet_batches_by_accounts.rs#L154 will use very few of these transactions.


Like if if had 250k packets in this buffer, we might only need 5k of them to fill all the buckets. We can probably leverage this to avoid cloning the entire buffer by popping only the ones we need, and then readding them back to the main data structure later

Reply:
Should try benching this with varying heap sizes to see what the perf looks like comparatively

Reply:
Did bunch of benching tests on `bankign_stage::filter_valid_packets_for_forwarding`, which includes clone/drain_desc MinMaxHeap, and sorting into buckets. With 10K packets/sec forwarding limitation in place, tested buffer of 10k forwardable packets, and 64k forwardable packets, found in both case the `filter_valid_packets_for_forwarding` took about same time (41ms, 39ms respectively). And the differences between drain_desc/collect and clone/drain_desc are also insignificant. I kept the clone/drain impl anyway, because it looks cleaner.

Reply:
Thanks for running that test, was this tested with the default 100 buckets?